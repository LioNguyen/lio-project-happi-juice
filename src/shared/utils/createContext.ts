import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
