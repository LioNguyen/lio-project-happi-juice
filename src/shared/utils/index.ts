export * from './lazyLoad'
export * from './tailwind'
