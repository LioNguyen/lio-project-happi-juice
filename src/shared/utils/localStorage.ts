/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
