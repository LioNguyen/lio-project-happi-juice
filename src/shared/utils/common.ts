import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
