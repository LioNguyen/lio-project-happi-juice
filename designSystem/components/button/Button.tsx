import { Slot } from "@radix-ui/react-slot";
import { cva, type VariantProps } from "class-variance-authority";
import * as React from "react";

import { cn } from "@/shared/utils";

const buttonVariants = cva(
  "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none  disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0",
  {
    variants: {
      variant: {
        default:
          "border-none bg-button text-button-foreground hover:bg-button/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90",
        destructive:
          "border-none bg-error text-button-foreground hover:bg-error/90 dark:bg-error dark:text-button-foreground dark:hover:bg-error/90",
        secondary:
          "border-none bg-button-secondary text-button-foreground shadow-sm hover:bg-button-secondary/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80",
        outline: "border border-input bg-none shadow-sm hover:bg-accent hover:text-accent-foreground",
        ghost: "hover:bg-accent hover:text-accent-foreground",
        link: "text-primary underline-offset-4 hover:underline",
      },
      size: {
        default: "h-9 px-4 py-2",
        sm: "h-8 rounded-md px-3 text-xs",
        lg: "h-10 rounded-md px-8",
        icon: "h-9 w-9",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "default",
    },
  },
);

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement>, VariantProps<typeof buttonVariants> {
  asChild?: boolean;
  isLoading?: boolean;
}

const LoadingSpinner = () => {
  return (
    <svg
      data-testid="loading-spinner"
      className="animate-spin"
      xmlns="http://www.w3.org/2000/svg"
      width="16"
      height="16"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    >
      <path d="M21 12a9 9 0 1 1-6.219-8.56" />
    </svg>
  );
};

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, isLoading = false, disabled, children, ...props }, ref) => {
    const Comp = asChild ? Slot : "button";
    return (
      <Comp
        data-testid="button-test"
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        disabled={disabled || isLoading}
        {...props}
      >
        {isLoading ? <LoadingSpinner /> : children}
      </Comp>
    );
  },
);
Button.displayName = "Button";

export default Button;
export { buttonVariants };
export type { ButtonProps };
