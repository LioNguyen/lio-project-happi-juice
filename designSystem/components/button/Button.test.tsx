import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { VariantProps } from "class-variance-authority";
import { ComponentPropsWithRef, createRef } from "react";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import Button, { buttonVariants } from "./Button";

type ButtonPropsWithRef = ComponentPropsWithRef<typeof Button>;
type ButtonVariantProps = VariantProps<typeof buttonVariants>;
type ButtonVariant = NonNullable<ButtonVariantProps["variant"]>;
type ButtonSize = NonNullable<ButtonVariantProps["size"]>;

const defaultProps: Partial<ButtonPropsWithRef> = {
  size: "default",
  variant: "default",
};

const renderComponent = (props: Partial<ButtonPropsWithRef> = {}) => {
  return renderWithTheme(<Button {...defaultProps} {...props} />);
};

describe("Button Component", () => {
  const defaultClass =
    "inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0";

  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  // Test render
  it("should render button with default props", () => {
    renderComponent({ children: "Click me" });

    const button = screen.getByRole("button", { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass(defaultClass);
  });

  // Test className custom
  it("should merge custom className with default classes", () => {
    renderComponent({ className: "custom-class" });

    const button = screen.getByRole("button");
    expect(button).toHaveClass("custom-class");
    expect(button).toHaveClass(defaultClass);
  });

  // Test asChild prop
  it("should render as child component when asChild is true", () => {
    renderComponent({ asChild: true, children: <a href="#">Link Button</a> });

    const link = screen.getByRole("link");
    expect(link).toBeInTheDocument();
    expect(link).toHaveClass(defaultClass);
  });

  // Test HTML attributes
  it("should forward HTML attributes", () => {
    renderComponent({ type: "submit", "aria-label": "Submit form" });

    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("type", "submit");
    expect(button).toHaveAttribute("aria-label", "Submit form");
  });

  // Test disabled state
  it("should be disabled when disabled prop is true", () => {
    renderComponent({ disabled: true });

    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
    expect(button).toHaveClass("disabled:pointer-events-none");
  });

  // Test ref forwarding
  it("should forward ref correctly", () => {
    const ref = createRef<HTMLButtonElement>();
    renderComponent({ ref });

    const button = screen.getByRole("button");
    expect(ref.current).toBe(button);
  });

  // Test variants
  describe("Variants", () => {
    const variants: ButtonVariant[] = ["default", "secondary", "outline", "ghost", "link"] as const;

    variants.forEach(variant => {
      it(`should render ${variant} variant correctly`, () => {
        renderComponent({ variant });

        const button = screen.getByRole("button");
        switch (variant) {
          case "default":
            expect(button).toHaveClass(defaultClass);
            break;
          case "secondary":
            expect(button).toHaveClass(
              "border-none bg-button-secondary text-button-foreground shadow-sm hover:bg-button-secondary/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80",
            );
            break;
          case "outline":
            expect(button).toHaveClass(
              "border border-input bg-none shadow-sm hover:bg-accent hover:text-accent-foreground",
            );
            break;
          case "ghost":
            expect(button).toHaveClass("hover:bg-accent hover:text-accent-foreground");
            break;
          case "link":
            expect(button).toHaveClass("text-primary underline-offset-4 hover:underline");
            break;
        }
      });
    });
  });

  // Test sizes
  describe("Sizes", () => {
    const sizes: ButtonSize[] = ["default", "sm", "lg", "icon"] as const;

    sizes.forEach(size => {
      it(`should render ${size} size correctly`, () => {
        renderComponent({ size });

        const button = screen.getByRole("button");
        switch (size) {
          case "default":
            expect(button).toHaveClass("h-9 px-4 py-2");
            break;
          case "sm":
            expect(button).toHaveClass("h-8 rounded-md px-3 text-xs");
            break;
          case "lg":
            expect(button).toHaveClass("h-10 rounded-md px-8");
            break;
          case "icon":
            expect(button).toHaveClass("h-9 w-9");
            break;
        }
      });
    });
  });

  // Test event handlers
  describe("Event handlers", () => {
    it("should handle click events", async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();

      renderComponent({ onClick: handleClick });

      const button = screen.getByRole("button");
      await user.click(button);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  // Test cho loading state
  describe("Loading State", () => {
    it("should show spinner and hide children when isLoading is true", () => {
      renderComponent({
        children: "Click me",
        isLoading: true,
      });

      const button = screen.getByRole("button");
      const spinner = screen.queryByTestId("loading-spinner");
      const text = screen.queryByText("Click me");

      expect(button).toBeDisabled();
      expect(spinner).toBeInTheDocument();
      expect(text).not.toBeInTheDocument();
    });

    it("should not show spinner and show children when isLoading is false", () => {
      renderComponent({
        children: "Click me",
        isLoading: false,
      });

      const button = screen.getByRole("button");
      const spinner = screen.queryByTestId("loading-spinner");
      const text = screen.queryByText("Click me");

      expect(button).not.toBeDisabled();
      expect(spinner).not.toBeInTheDocument();
      expect(text).toBeInTheDocument();
    });

    it("should disable button and prevent click events when loading", async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();

      renderComponent({
        onClick: handleClick,
        isLoading: true,
        children: "Click me",
      });

      const button = screen.getByRole("button");

      expect(button).toBeDisabled();
      await user.click(button);
      expect(handleClick).not.toHaveBeenCalled();
    });

    it("should maintain button variants when loading", () => {
      renderComponent({
        variant: "secondary",
        isLoading: true,
        children: "Click me",
      });

      const button = screen.getByRole("button");
      expect(button).toHaveClass("bg-button-secondary");
    });
  });
});
