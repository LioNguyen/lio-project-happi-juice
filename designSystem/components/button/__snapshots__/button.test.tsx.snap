// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > should match snapshot with default props 1`] = `
<button
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border-none bg-button text-button-foreground hover:bg-button/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90 h-9 px-4 py-2"
  data-testid="button-test"
/>
`;
